### PR TITLE
156 solve like function which dispatches on epiawareproblem and epiawaremethod

### LINKED ISSUE #120: Passing initialisations to NUTS/inference approach
Can we have a helper function for this or make it part of the output of best_pf?  I don't love reading this or the thought of having to repeatedly write it down.

_Originally posted by @seabbs in https://github.com/CDCgov/Rt-without-renewal/pull/117#discussion_r1516011356_

### LINKED ISSUE #156: `solve`- like function which dispatches on `EpiAwareProblem` and `EpiAwareMethod`

### PR DESCRIPTION
**Pull request by SamuelBrand1**
A draft PR for an `apply_method` method.

## Methods
The basic idea is that each behaviour of a `method`, which means either a sampling or optimization technique, has an `_apply_method` dispatch which implements this.

An important method is `EpiMethod` implies doing a sequence of optimizations passing infomation to each next level, then finally to a sampling method `NUTSampler`.

## `apply_method`

`apply_method` has two methods:
1. ingests a `Turing` model object and applies the epi method.
2. Ingests an `EpiProblem` and a data object, as well parameters to fix and/or condition on, creates an appropriate model and then apply the epi method.

The output of `apply_method` is a `NamedTuple` which returns `samples` as well as `gens`, that is generated quantities if that makes sense for the returned object from `_apply_method`, and `model` which is the model (after any conditioning or fixing of parameters).

## TBD

- Naming?
- Return objects? 
- More unit tests

### CODE REVIEW

File: EpiAware/src/EpiInference/EpiInference.jl [hunk 1]
```diff
@@ -1,17 +1,18 @@
...
 """
 module EpiInference
 
-using ..EpiAwareBase: AbstractEpiMethod, AbstractEpiOptMethod,
-                      AbstractEpiSamplingMethod
-using Pathfinder: pathfinder, PathfinderResult
-
-using DynamicPPL, DocStringExtensions
+using ..EpiAwareBase: AbstractEpiMethod, AbstractEpiOptMethod, AbstractEpiSamplingMethod
+import ..EpiAwareBase: _apply_method
```
Comment (seabbs) [inline EpiAware/src/EpiInference/EpiInference.jl:7]: why are you using both here? I thought we decided only to use using?

**Review by seabbs — COMMENTED** [review]
In general I think this all looks good. Can look through in more detail or happy to wait until it goes from draft to full.

File: EpiAware/src/EpiInference/EpiInference.jl (hunk 1 above)
Comment (SamuelBrand1) [reply EpiAware/src/EpiInference/EpiInference.jl:7]: `import`ing a function means that each method on `_apply_method` is associated with that function. `using` only does that if you prefix the method definition with the module. I'm happy to do that instead if you prefer.

Here I wanted `_apply_method` in `EpiAwareBase` because its used differently in the inference submodule and the infection submodule.

**Comment by SamuelBrand1** [comment]
Its failing the 
> Testing NegativeBinomialError against theoretical properties` 
test on CI but not locally. Any thoughts?

**Comment by seabbs** [comment]
I think this is ready for review/merging now. 

Whilst there are definitely some edge case issues and potentially testing gaps I think we would be best served in making new issues for those based on review and only fixing large scale issues before merging (as this is already very large).

**Comment by seabbs** [comment]
@SamuelBrand1 this is ready for review I think. 

There is an outstanding issue with the manypathfinder into NUTs in the getting started example that I think is unrelated but I am really not clear on. 

We also need to open some new issues to improve areas of this but as long as we do that I think we should move to merge asap as this is now very large, blocking, and complicated (none of which is ideal).

**Comment by SamuelBrand1** [comment]
OK. I'm looking at this and the getting_started example

**Review by SamuelBrand1 — COMMENTED** [review]
First, quick takeaway. I like the restructure in the files.

File: EpiAware/src/EpiAwareBase/EpiAwareObservables.jl [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+- `data`: The data used for the observables.
+- `samples`: Samples from the posterior distribution.
+- `generated`: The generated observables.
+"
+struct EpiAwareObservables
+    model::Any
+    data::Any
+    samples::Any
+    generated::Any
+end
```
Comment (SamuelBrand1) [inline EpiAware/src/EpiAwareBase/EpiAwareObservables.jl:15]: I like this.

File: EpiAware/src/EpiAwareBase/types.jl [hunk 3]
```diff
@@ -1,17 +1,27 @@
 abstract type AbstractModel end
 
 """
-The abstract supertype for all structs that define a model for generating unobserved/latent
-infections.
+The abstract supertype for all structs that define a model for generating
+unobserved/latent infections.
 """
```
Comment (SamuelBrand1) [inline EpiAware/src/EpiAwareBase/types.jl:6]: Not really a problem, but what config are we going for on doc string cols allowed per line?

File: EpiAware/src/EpiInfModels/EpiData.jl [hunk 4]
```diff
@@ -1,4 +1,4 @@
-"""
+@doc raw"
```
Comment (SamuelBrand1) [inline EpiAware/src/EpiInfModels/EpiData.jl:1]: Is there any use-case for _not_ using `@doc raw"...some docs..."` pattern?

**Review by SamuelBrand1 — COMMENTED** [review]
Looking at `getting_started` now

File: EpiAware/src/EpiInfModels/EpiData.jl (hunk 4 above)
Comment (seabbs) [reply EpiAware/src/EpiInfModels/EpiData.jl:1]: I don't think so?

File: EpiAware/docs/src/examples/getting_started.jl [hunk 5]
```diff
@@ -174,39 +190,85 @@ y_t \sim \text{NegBinomial}\Big(\mu = \sum_{s\geq 0} K[t, t-s] I(s), r\Big). \\
...
+    tspan = (1, 100))
+
 # ╔═╡ e813d547-6100-4c43-b84c-8cebe306bda8
 md"
 We also set up the inference to occur over 100 days.
 "
 
 # ╔═╡ c7580ae6-0db5-448e-8b20-4dd6fcdb1ae0
-time_horizon = 100
+time_horizon = 30
```
[the remark this answers is not in the document]
Comment (SamuelBrand1) [reply EpiAware/docs/src/examples/getting_started.jl:217]: Yes. I think we have a few issues on the tutorial:

- What workflow is the best to see first.
- Most representative model to use.

File: EpiAware/src/EpiAwareBase/types.jl (hunk 3 above)
Comment (SamuelBrand1) [reply EpiAware/src/EpiAwareBase/types.jl:6]: Shall we make that a standard. I was doing 92 char per line for doc strings.

File: EpiAware/docs/src/examples/getting_started.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (seabbs) [reply EpiAware/docs/src/examples/getting_started.jl]: So those are issues with the tutorial on `main` IMO. Here we are just trying to get it to work as well as the one on main (which is working not that well but doesn't fail)

**Comment by SamuelBrand1** [comment]
Want to make me a reviewer? I'm distant enough I think due to going on holiday in the middle of this PR?

**Comment by seabbs** [comment]
I can't becaise you opened it

**Comment by SamuelBrand1** [comment]
> I can't because you opened it. We need to just mutually agree we are okay with merging and then can do

I'm good. I approve.

**Review by seabbs — APPROVED** [review]
